### PR TITLE
Add support for Nexys Video board

### PR DESCRIPTION
**Pull request by Saute0212**
I referred to #350.
I confirmed that I could boot Linux on the Nexys Video board.
I added the bootrom generation process to the Makefile, because nexys video needs custom parameters in dts file.
This work was implemented in the Nakajo Laboratory at Tokyo University of Agriculture and Technology.

### CODE REVIEW

**Comment by JeanRochCoulon** [comment]
Great contribution !   @jquevremont can you review the PR ?

**Comment by jquevremont** [comment]
@Saute0212 You need to sign Eclipse ECA to submit PRs. Then, you should make sure you use the same email address in your GitHub and Eclipse accounts.

**Comment by Saute0212** [comment]
Thanks for the reply.
I have signed the Eclipse ECA.
I signed the commit with ```commit -s```.

Also I am checking Genesys2 support.

**Comment by Saute0212** [comment]
@jquevremont 
I have confirmed that I can boot Linux on Genesys2 without any problems.

**Comment by JeanRochCoulon** [comment]
Hello, as you can see in the CI, the cv32a60x FPGA build fails on GenesisII.

**Comment by Saute0212** [comment]
@JeanRochCoulon 
I am unable to view the logs and therefore cannot respond to the error.

**Comment by Saute0212** [comment]
Sorry. I do not know why the "Environment check" and "FPGA Build cv32a60x" tests are failing. I can't see the error logs so I cannot respond to the error.

**Comment by JeanRochCoulon** [comment]
@Gchauvon Can you provide information to @Saute0212 to help to debug GenesysII build fail.

File: corev_apu/fpga/src/bootrom/Makefile [hunk 1]
```diff
@@ -1,11 +1,31 @@
 XLEN ?= 64
 CROSSCOMPILE ?= riscv64-unknown-elf-
```
Comment (Gchauvon) [inline corev_apu/fpga/src/bootrom/Makefile:2]:
```suggestion
CROSSCOMPILE ?= riscv-none-elf-
```

File: corev_apu/fpga/src/bootrom/Makefile [hunk 2]
```diff
@@ -1,11 +1,31 @@
...
+endif
+
+CC = $(RISCV)/bin/${CROSSCOMPILE}gcc
+OBJCOPY = $(RISCV)/bin/$(CROSSCOMPILE)objcopy
+SED = sed
+PLATFORM_DEFINES = -DCLOCK_FREQUENCY=$(CLOCK_FREQUENCY) -DUART_BITRATE=$(UART_BITRATE)
 
 ifeq ($(XLEN), 64)
-CFLAGS = -Os -ggdb -march=rv64imac -mabi=lp64 -Wall -mcmodel=medany -mexplicit-relocs
+CFLAGS = $(PLATFORM_DEFINES) -Os -ggdb -march=rv64imac -mabi=lp64 -Wall -mcmodel=medany -mexplicit-relocs -ffreestanding
```
Comment (Gchauvon) [inline corev_apu/fpga/src/bootrom/Makefile:26]:
```suggestion
CFLAGS = $(PLATFORM_DEFINES) -Os -ggdb -march=rv64gc_zba_zbb_zbs_zbc_zicsr_zifencei -mabi=lp64 -Wall -mcmodel=medany -mexplicit-relocs -ffreestanding
```

File: corev_apu/fpga/src/bootrom/Makefile [hunk 3]
```diff
@@ -1,11 +1,31 @@
...
 else
-CFLAGS = -O2 -ggdb  -march=rv32im -mabi=ilp32 -W -Wall -mcmodel=medany -mexplicit-relocs -fno-builtin
+CFLAGS = $(PLATFORM_DEFINES) -O2 -ggdb  -march=rv32im -mabi=ilp32 -W -Wall -mcmodel=medany -mexplicit-relocs -fno-builtin -ffreestanding
```
Comment (Gchauvon) [inline corev_apu/fpga/src/bootrom/Makefile:28]:
```suggestion
CFLAGS = $(PLATFORM_DEFINES) -O2 -ggdb  -march=rv32imac_zba_zbb_zbs_zbc_zicsr_zifencei -mabi=ilp32 -W -Wall -mcmodel=medany -mexplicit-relocs -fno-builtin -ffreestanding
```

**Review by Gchauvon — CHANGES_REQUESTED** [review]
Old bootrom_32 and bootrom_64 are removed. It needs to be recompiled but the Makefile to do so was not compatible with gcc13.1.0 which is the version required for cva6.

I don't know what impact `-ffreestanding` will have on the generated bootrom.

**Comment by Saute0212** [comment]
Thanks for the reply.
I will try and test your suggestion.
Thank you very much.

**Comment by Saute0212** [comment]
As you can see, the "environment check" still fails.
Also, the cross-compiler option was changed to `rv64imac_zba_zbb_zbs_zbc_zicsr_zifencei` instead of `rv64gc_zba_zbb_zbs_zbc_zicsr_zifencei`. 
Because bootrom did not work correctly when the cross-compiler option was `rv64gc_zba_zbb_zbs_zbc_zicsr_zifencei`.

**Comment by Saute0212** [comment]
@Gchauvon 
As you can see, "Environment check" is failed.
When the cross-compiler option is ```rv64imac_zba_zbb_zbs_zbc_zicsr_zifencei```, Linux can be booted.
But when the cross-compiler option is ```rv64gc_zba_zbb_zbs_zbc_zicsr_zifencei```, the following output is displayed and the bootrom does not work correctly.
What corrections are needed?
```
Hello World!
init SPI
status: 0x0000000000000025
status: 0x0000000000000025
SPI initialized!
initializing SD...
SD command cmd0           response : 01
SD command cmd55         response : 01
SD command cmd41         response : 01
SD command cmd55         response : 01
SD command cmd41         response : 00
sd initialized!
```

**Comment by Saute0212** [comment]
@Gchauvon 
Sorry to ping you again, but "Environment check" is still failing, and I have no way of checking the logs. Is it possible for you to show me the logs, or help me debug?

When the cross-compiler option is ```rv64imac_zba_zbb_zbs_zbc_zicsr_zifencei```, Linux can be booted.
But when the cross-compiler option is ```rv64gc_zba_zbb_zbs_zbc_zicsr_zifencei```, the following output is displayed and the bootrom does not work correctly.
```
Hello World!
init SPI
status: 0x0000000000000025
status: 0x0000000000000025
SPI initialized!
initializing SD...
SD command cmd0           response : 01
SD command cmd55         response : 01
SD command cmd41         response : 01
SD command cmd55         response : 01
SD command cmd41         response : 00
sd initialized!
```

**Review by Gchauvon — APPROVED** [review]
Environment check failure is not due to this PR. LGTM

**Comment by amiroshni** [comment]
@Gchauvon , thank you for your response and for pointing me to the SDK repo.

I was able to follow the instructions to compile an image, as well as setup the SD card for the Genesys 2.
The fw_payload.bin and uImage were copied over to SD card with 'make flash-sdcard' command.
Didn't expect it to work with the Nexys Video, but tried anyway, and got the following:

```
init SPI
status: 0x0000000000000025
status: 0x0000000000000025
SPI initialized!
initializing SD... 
SD command cmd0 	response : 01
SD command cmd55 	response : 01
SD command cmd41 	response : 01
[...skipping for brevity...]
SD command cmd55 	response : 01
SD command cmd41 	response : FF
could not initialize sd... exiting
```

@Saute0212 , was able to get Linux running, as mentioned in his first message.
Are there commits available which make the required changes to the CVA6 SDK repo to support the Nexys Video board?

**Comment by Saute0212** [comment]
@amiroshni 
I'm glad you implemented CVA6 on the Nexys Video board.

Nexys Video cannot use Ethernet.
Please execute the following command in the `cva6-sdk` directory.
```
sed -i 's/BR2_PACKAGE_ETHTOOL=y/BR2_PACKAGE_ETHTOOL=n/' configs/buildroot64_defconfig
sed -i -e 's/CONFIG_NET=y/CONFIG_NET=n/' -e 's/CONFIG_INET=y/CONFIG_INET=n/' -e 's/CONFIG_NETDEVICES=y/CONFIG_NETDEVICES=n/' -e 's/CONFIG_LOWRISC_DIGILENT_100MHZ=y/CONFIG_LOWRISC_DIGILENT_100MHZ=n/' configs/linux64_defconfig
make images
```

After `make images` finishes running, write the [URL](https://github.com/openhwgroup/cva6-sdk/releases/download/v0.3.0/bbl.bin) data (`bbl.bin`) to the first partition of the microSD card.
In the second partition, write the `uImage` generated by make images.

Please contact me if you have any questions.

**Comment by amiroshni** [comment]
Thank you @Saute0212 !

I disabled network controller as you instructed and generated a new uImage.

Downloaded the bbl.bin file (I guess it's equivalent to the original fw_payload.bin, but much bigger?), and made several changes to the Makefile for easier flashing. The changes are kind of hacky, so I wouldn't PR them until someone has a look at it, but my changes are available in a branch:
https://github.com/openhwgroup/cva6-sdk/commit/bb557e2e995e4598ba9c280e7600357893930762

I didn't add the wget instruction to the makefile yet, so has to be run separately (it is assumed cva6-sdk is in user's home):

```
wget -nc -P $HOME/cva6-sdk/install64/ https://github.com/openhwgroup/cva6-sdk/releases/download/v0.3.0/bbl.bin
sudo make flash-sdcard PAYLOAD=bbl.bin SDDEVICE=/dev/[YOUR SD CARD DEVICE] RISCV=$HOME/cva6-sdk/install64
```

Flashing worked, testing on the FPGA, I get the following serial message:

```
init SPI
status: 0x0000000000000025
status: 0x0000000000000025
SPI initialized!
initializing SD... 
SD command cmd0 	response : 01
SD command cmd55 	response : 01
SD command cmd41 	response : 01
SD command cmd55 	response : 01
SD command cmd41 	response : 01
SD command cmd55 	response : 01
SD command cmd41 	response : 00
sd initialized!
```

Not sure if I've missed something else, looks like nothing is actually booting.

**Comment by Saute0212** [comment]
@amiroshni 
I would like to debug, so please tell me about the following.

- Cross compiler version
- Capacity of microSD card
- Revision number of the commit in the CVA6 repository you used

**Comment by amiroshni** [comment]
@Saute0212 

Sure, thank you for spending the time on this.

- Cross compiler... do mean the compiler used in the cva6sdk or the cva6 repo?
  - Version setup with RISC-V toolchain using [script](https://github.com/openhwgroup/cva6/blob/bfff84eaebcc46d07a9274419216e3fb6508f567/util/gcc-toolchain-builder/get-toolchain.sh) in cva6 repo: riscv-none-elf-gcc (GCC) 13.1.0
  - HOSTCC used during cva6sdk 'make images': gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
  - CROSS_COMPILE used during cva6sdk 'make images': riscv64-buildroot-linux-gnu-gcc.br_real (Buildroot 2021.08) 10.3.0

- Capacity of microSD card: SanDisk Ultra 16GB (15.9GB/14.8GiB)

- CVA6 repo commit: https://github.com/openhwgroup/cva6/commit/bfff84eaebcc46d07a9274419216e3fb6508f567

Let me know if I've missed something, or if more information is needed.

**Comment by Saute0212** [comment]
@amiroshni 
Thanks for the reply.
Please wait a while as I try to implement it.

**Comment by Saute0212** [comment]
@amiroshni 
Sorry for the late reply.

I changed ```-march=rv64imac_zba_zbb_zbs_zbc_zicsr_zifencei``` to ```-march=rv64imac_zicsr``` in ```cva6/corev_apu/fpga/src/bootrom/Makefile```.
Then I was able to boot Linux.

**Comment by amiroshni** [comment]
@Saute0212 , thank you for checking your configurations.

My colleague and I made the `-march` flag change in the CVA6's bootrom, and re-generated the bitstream.

However Linux is still not booting from SD card. This is the current print out:

```
Hello from RED!
init SPI
status: 0x0000000000000025
status: 0x0000000000000025
SPI initialized!
initializing SD...
SD command cmd0 response : 01
SD command cmd55 response : 01
SD command cmd41 response : 01
SD command cmd55 response : 01
SD command cmd41 response : 01
SD command cmd55 response : 01
SD command cmd41 response : 00
sd initialized!
```

The full instructions for what we did can be found here:
https://github.com/amiroshni/documenting-cva6

Are you sure that we're not using a different version of the bootroom?
Our bootrom version is at CVA6 commit: https://github.com/openhwgroup/cva6/commit/bfff84eaebcc46d07a9274419216e3fb6508f567
Our cva6-sdk version: https://github.com/amiroshni/cva6-sdk/commit/bb557e2e995e4598ba9c280e7600357893930762
(my fork, on the nexys-video-test branch)

**Comment by Saute0212** [comment]
@amiroshni 
Thanks for contacting me.
I have confirmed that the version of bootrom is correct.

I have three questions.
- Did you run ```make fpga``` after running ```make clean```?
- Is it possible to reduce the capacity of the microSD card? For example, 4GB.
- Could you try to clone the repository I am forking and generate Bitstream?
CVA6 repository I forked : https://github.com/Saute0212/cva6

Thank you.

**Comment by amiroshni** [comment]
Thank you for the fast response @Saute0212 !

My colleague and I followed your suggestions (except for the 4GB microsd card, but we ordered some).

> @amiroshni Thanks for contacting me. I have confirmed that the version of bootrom is correct.
> 
> I have three questions.
> 
>     * Did you run `make fpga` after running `make clean`?
Yes, we did.

> 
>     * Is it possible to reduce the capacity of the microSD card? For example, 4GB.
Currently we only have the 16GB cards on hand, but ordered some 4GB ones for testing. Should arrive in 3-5 days.

> 
>     * Could you try to clone the repository I am forking and generate Bitstream?
>       CVA6 repository I forked : https://github.com/Saute0212/cva6
Yes. Changed my documentation to use your cva6 fork: https://github.com/amiroshni/documenting-cva6

My colleague @shriyasharma11 generated the new bitstream and created a serial log of the boot process here: https://gist.github.com/shriyasharma11/12136b710edaa7eac2e1269a903f8cef

Interesting problems seem to repeat throughout the kernel boot:
```
[   39.968943] mmc0: new SDHC card on SPI
[   40.105682] mmcblk0: mmc0:0000 SC16G 14.8 GiB 
Starting logging: [   40.380820]  mmcblk0: unable to read partition table
[   40.419800] mmc0: SPI card removed
[   41.148149] mmc0: host does not support reading read-only switch, assuming write-enable
[   41.170358] mmc0: new SDHC card on SPI
[   41.380054] mmcblk0: mmc0:0000 SC16G 14.8 GiB 
[   41.731772]  mmcblk0: unable to read partition table
[   41.835341] mmc0: SPI card removed
OK
[   42.674518] mmc0: error -110 whilst initialising SD card
```
(even though the micro sd card is not touched, kernel seems to think it's being removed).

We will also try on a different 16GB card (we have three of those), and once the 4GB arrives will test as well.

**Comment by AngelaGonzalezMarino** [comment]
Hi @Saute0212 @JeanRochCoulon 

I have realised that after this PR the bitstream generated for cv32a6_ima_sv32_fpga does not boot Linux.
I tried on Genesys 2, but I think it will be the same for Nexys Video

The issue is that the makefile for bootrom generation now uses 
-march=rv32imac_zba_zbb_zbs_zbc_zicsr_zifencei which assumes that a few extensions are always there

In the case of cv32a6_ima_sv32_fpga the C extension (and others) is not there, therefore it never boots

Given that extensions are optional, I think we should keep the simplest march, to have a common bootrom compilation and avoid this kind of problem. What do you think?

I have been able to successfully generate a bitstream and boot correctly on Genesys2 with the following configurations:
- cv32a6_ima_sv32_fpga using -march=rv32im_zicsr 
- cv64a6_imafdch_sv39   using -march=rv64im_zicsr

**Comment by kkhalfallah** [comment]
Hi everyone,
Thanks a lot for that work. Following this material I also manage to boot Linux on the nexys_video board.
Do we have an idea why the ethernet can't work on the nexys_video board, while it can on the Genesys 2?
This is weird because the two boards are populated with the very same Ethernet PHY device (namely the Realtek RTL8211E-VL) which can be

**Comment by kkhalfallah** [comment]
Hi everyone,
Thanks a lot for that work. Following this material I also manage to boot Linux on the nexys_video board.
Do we have an idea why the ethernet can't work on the nexys_video board, while it can on the Genesys 2?
This is weird because the two boards are populated with the very same Ethernet PHY device (namely the Realtek RTL8211E-VL).
This can be checked here for the nexys_video board:
https://digilent.com/reference/programmable-logic/nexys-video/reference-manual
(search for string "RTL8211E")
and here for the Genesys-2 board:
https://digilent.com/reference/programmable-logic/genesys-2/reference-manual
(search for same string "RTL8211E").
Regards!
Thanks again.
Karim

**Comment by jquevremont** [comment]
Sorry @kkhalfallah not to be able to hrlp as we do not use this board.

Si @Saute0212, we have designed a "CVA6 Ecosystem and Resources" page. How could we mention the support of the Nexys board?

**Comment by kkhalfallah** [comment]
Hello @jquevremont,
Thanks for your reply. Difficult to mention anything about the nexys_video board if we still can't run a mere SSH on it, in my humble opinion.
I will try to make progress on the subject and if it's conclusive I will reach back to you.
Best regards and thanks again!
Karim

**Pull request merged**